content/match-types-spec.md

```diff
@@ -0,0 +1,354 @@
+Indeed, because of it, a pattern that is legal at definition site may become illegal after some later substitution.
+
+Consider for example the standard library's very own `Tuple.InverseMap`:
+
+```scala
+/** Converts a tuple `(F[T1], ..., F[Tn])` to `(T1,  ... Tn)` */
+type InverseMap[X <: Tuple, F[_]] <: Tuple = X match {
+  case F[x] *: t => x *: InverseMap[t, F]
+  case EmptyTuple => EmptyTuple
+}
+```
+
+If we instantiate `InverseMap` with a class type parameter, such as `InverseMap[X, List]`, the first case gets instantiated to
+```scala
+case List[x] *: t => x *: InverseMap[t, List]
+```
+which is legal.
+
+However, nothing prevents us a priori to instantiate `InverseMap` with an illegal type constructor, for example
+```scala
+type IsSeq[t <: Seq[Any]] = t
+InverseMap[X, IsSeq]
+```
+which gives
+```scala
+case IsSeq[x] *: t => x *: InverseMap[t, IsSeq]
+```
+
+These instantiatiations happen deep inside the type checker, during type computations.
+Since types are cached, shared and reused in several parts of the program, by construction, we do not have any source code position information at that point.
+That means that we cannot report *warnings*.
+
+We can in fact report *errors* by reducing to a so-called `ErrorType`, which is aggressively propagated.
+This is what we do in the proposed implementation (unless using `-source:3.3`).
+
+### Open questions
+
+None at this point.
+
+## Alternatives
+
+The specification is more complicated than we initially wanted.
+At the beginning, we were hoping that we could restrict match cases to class type constructors only.
+The quantitative study however revealed that we had to introduce support for abstract type constructors and for type member extractors.
+
+As already mentioned, the standard library itself contains an occurrence of an abstract type constructor in a pattern.
+Making that an error would mean declaring the standard library itself bankrupt, which was not a viable option.
+
+We tried to restrict abstract type constructor to never match on their own.
+Instead, we wanted them to stay *stuck* until they could be instantiated to a concrete type constructor.
+However, that led some existing tests to fail even for match types that were declared legal, because they did not reduce anymore in some places where they reduced before.
+
+Type member extractors are our biggest pain point.
+Their specification is complicated, and the implementation as well.
+Our quantitative study showed that they were however "often" used (10 occurrences spread over 4 libraries).
+In each case, they seem to be a way to express what Scala 2 type projections (`A#T`) could express.
+While not quite as powerful as type projections (which were shown to be unsound), match types with type member extractors delay things enough for actual use cases to be meaningful.
+
+As far as we know, those use cases have no workaround if we make type member extractors illegal.
```

Comment:
Hum no, the sentence says: if we remove type member extractors *from the proposed spec*, then we will have no workaround anymore. With the proposed spec, we can extract type members.

The originally submitted SIP/spec did not allow to extract the type member of `SomeEnumeration#Value` because `Value` is a *class* member (not a type alias), but the commit "Allow class members to be matched by type member extractors." made that actually possible.

Reply:
OK thanks for the clarification!